### PR TITLE
Stabilize `Result::as_deref` and `as_deref_mut`

### LINKED ISSUE #50264: Tracking issue for Option::deref, Result::deref, Result::deref_ok, and Result::deref_err (feature: inner_deref)
## API

```rust
impl<T, E> Result<T, E> {
    fn as_deref(&self) -> Result<&T::Target, &E> where T: Deref;
    fn as_deref_err(&self) -> Result<&T, &E::Target> where E: Deref;
    fn as_deref_mut(&mut self) -> Result<&mut T::Target, &mut E> where T: DerefMut;
    fn as_deref_mut_err(&mut self) -> Result<&mut T, &mut E::Target> where E: DerefMut;
}
```

<br>

## Implementation history

- Implemented in https://github.com/rust-lang/rust/pull/50267 as:

    ```rust
    impl<T> Option<T> {
        fn deref(&self) -> Option<&T::Target> where T: Deref;
    }
    impl<T, E> Result<T, E> {
        fn deref(&self) -> Result<&T::Target, &E::Target> where T: Deref, E: Deref;
        fn deref_ok(&self) -> Result<&T::Target, &E> where T: Deref;
        fn deref_err(&self) -> Result<&T, &E::Target> where E: Deref;
    }
    ```

- Renamed and extended in #59628 / #62421:

    ```rust
    impl<T> Option<T> {
        fn as_deref(&self) -> Option<&T::Target> where T: Deref;
        fn as_deref_mut(&mut self) -> Option<&mut T::Target> where T: DerefMut;
    }
    impl<T, E> Result<T, E> {
        fn as_deref(&self) -> Result<&T::Target, &E::Target> where T: Deref, E: Deref;
        fn as_deref_ok(&self) -> Result<&T::Target, &E> where T: Deref;
        fn as_deref_err(&self) -> Result<&T, &E::Target> where E: Deref;
        fn as_deref_mut(&mut self) -> Result<&mut T::Target, &mut E::Target> where T: DerefMut, E: DerefMut;
        fn as_deref_mut_ok(&mut self) -> Result<&mut T::Target, &mut E> where T: DerefMut;
        fn as_deref_mut_err(&mut self) -> Result<&mut T, &mut E::Target> where E: DerefMut;
    }
    ```

- `Option` methods stabilized in #64708:

    ```rust
    impl<T> Option<T> {
        fn as_deref(&self) -> Option<&T::Target> where T: Deref;
        fn as_deref_mut(&mut self) -> Option<&mut T::Target> where T: DerefMut;
    }
    ```

- `Result` methods pared down and renamed in #67930:

    ```rust
    impl<T, E> Result<T, E> {
        fn as_deref(&self) -> Result<&T::Target, &E> where T: Deref;
        fn as_deref_err(&self) -> Result<&T, &E::Target> where E: Deref;
        fn as_deref_mut(&mut self) -> Result<&mut T::Target, &mut E> where T: DerefMut;
        fn as_deref_mut_err(&mut self) -> Result<&mut T, &mut E::Target> where E: DerefMut;
    }
    ```

### PR DESCRIPTION
**Pull request by tesuji**
FCP completed in https://github.com/rust-lang/rust/issues/50264#issuecomment-645681400.

This PR stabilizes two methods for `std::result::Result`:
```rust
fn as_deref(&self) -> Result<&T::Target, &E> where T: Deref;
fn as_deref_mut(&mut self) -> Result<&mut T::Target, &mut E> where T: DerefMut;
```

This PR also removes two rarely used unstable methods from `Result`:
```rust
fn as_deref_err(&self) -> Result<&T, &E::Target> where E: Deref;
fn as_deref_mut_err(&mut self) -> Result<&mut T, &mut E::Target> where E: DerefMut;
```

Closes #50264

### CODE REVIEW

**Review by dtolnay — APPROVED** [review]
Thanks!

**Comment by dtolnay** [comment]
@bors r-

Libcore tests need to be updated.

```console
error[E0599]: no method named `as_deref_err` found for reference `&std::result::Result<u8, &i32>` in the current scope
   --> library/core/tests/result.rs:257:24
    |
257 |     assert_eq!(ref_err.as_deref_err(), expected_result);
    |                        ^^^^^^^^^^^^ method not found in `&std::result::Result<u8, &i32>`
```

File: library/core/src/result.rs [hunk 1]
```diff
@@ -1164,23 +1162,12 @@ impl<T: Deref, E> Result<T, E> {
     /// let y: Result<&str, &u32> = Err(&42);
     /// assert_eq!(x.as_deref(), y);
     /// ```
+    #[stable(feature = "inner_deref", since = "1.47.0")]
```
Comment (lzutao) [inline library/core/src/result.rs:1165]: About feature gate, do you in favor of renaming it to "result_as_deref" or so?

**Review by dtolnay — APPROVED** [review]
Thanks!

File: library/core/src/result.rs (hunk 1 above)
Comment (dtolnay) [reply library/core/src/result.rs:1165]: I would just leave it I think.

**Pull request merged**